pkg/bgpv1/manager/reconcilerv2/paths.go

```diff
@@ -152,34 +153,76 @@ func reconcilePaths(params *reconcilePathsParams) (PathMap, error) {
 
 	// withdraw unneeded adverts
 	for advrtKey, advrt := range toWithdraw {
-		l.WithFields(logrus.Fields{
-			types.PathLogField:   advrt.NLRI.String(),
-			types.FamilyLogField: advrt.Family.String(),
-		}).Debug("Withdrawing path")
-
-		if err := params.Router.WithdrawPath(params.Ctx, types.PathRequest{Path: advrt}); err != nil {
+		err := withdrawPath(params, runningAdverts, advrtKey, advrt)
+		if err != nil {
 			return runningAdverts, err
 		}
-		delete(runningAdverts, advrtKey)
 	}
 
 	// create new adverts
 	for advrtKey, advrt := range toAdvertise {
-		l.WithFields(logrus.Fields{
-			types.PathLogField:   advrt.NLRI.String(),
-			types.FamilyLogField: advrt.Family.String(),
-		}).Debug("Advertising path")
-
-		advrtResp, err := params.Router.AdvertisePath(params.Ctx, types.PathRequest{Path: advrt})
+		err := advertisePath(params, runningAdverts, advrtKey, advrt)
 		if err != nil {
 			return runningAdverts, err
 		}
-		runningAdverts[advrtKey] = advrtResp.Path
 	}
 
 	return runningAdverts, nil
 }
 
+func advertisePath(params *reconcilePathsParams, runningAdverts PathMap, pathKey string, path *types.Path) error {
+	if params.PathReferences != nil && params.PathReferences[pathKey] > 0 {
+		// path already advertised for another resource
+		params.PathReferences[pathKey] += 1
+		runningAdverts[pathKey] = path
+		return nil
+	}
+
+	params.Logger.WithFields(logrus.Fields{
+		types.PathLogField:   path.NLRI.String(),
+		types.FamilyLogField: path.Family.String(),
+	}).Debug("Advertising path")
+
+	advrtResp, err := params.Router.AdvertisePath(params.Ctx, types.PathRequest{Path: path})
+	if err != nil {
+		return err
+	}
+
+	// update only in case of no error
+	if params.PathReferences != nil {
+		params.PathReferences[pathKey] += 1
+	}
+	runningAdverts[pathKey] = advrtResp.Path
+
+	return nil
+}
+
+func withdrawPath(params *reconcilePathsParams, runningAdverts PathMap, pathKey string, path *types.Path) error {
+	if params.PathReferences != nil && params.PathReferences[pathKey] > 1 {
+		// path still needs to be advertised for another resource
+		params.PathReferences[pathKey] -= 1
+		delete(runningAdverts, pathKey)
+		return nil
+	}
+
+	params.Logger.WithFields(logrus.Fields{
+		types.PathLogField:   path.NLRI.String(),
+		types.FamilyLogField: path.Family.String(),
+	}).Debug("Withdrawing path")
+
+	if err := params.Router.WithdrawPath(params.Ctx, types.PathRequest{Path: path}); err != nil {
+		return err
+	}
+
+	// update only in case of no error
+	if params.PathReferences != nil {
+		params.PathReferences[pathKey] -= 1
+	}
+	delete(runningAdverts, pathKey)
+
+	return nil
+}
+
```

Comment:
Do we need to implement this logic inside the path reconciler? We can instead let Service reconciler to calculate the desired paths with reference counting? So that we don't need to modify path reconciler.

Reply:
To me Paths reconciler is a bit more natural place for this: firstly because we may reuse this in other reconcilers in the future if needed. Secondly, if we moved this logic to the service reconciler before the Path reconciler is called, I think the logic would need to be a bit more complex (and potentially more error prone) than just guarding AdvertisePath / WithdrawPath calls - for example we would need to manipulate [desiredPaths / currentPaths in reconcilePaths](https://github.com/cilium/cilium/blob/4f00d67cb4a4f44c487f9308e0184815ab86334b/pkg/bgpv1/manager/reconcilerv2/service.go#L317-L318). But it's doable, WDYT about this @harsimran-pabla ?

Reply:
Yeah, I see what you mean. Probably ideal would be if we could move this completely into the Path "reconciler" (`ReconcileAFPaths` func execution). The problem is that currently it is executed per-resource (service, pool, etc.), so we would need to change it to work with the set of all resources - I think that should possible, let me see how big change that would be.

Reply:
OK, I think we should do it this way - implemented a new Path reconciler func `ReconcileResourceAFPaths` in https://github.com/cilium/cilium/pull/35166/commits/05067398a8bce8717e3748382c154aa0d970f4c7 (can squash the two commits before merging if we are good with this approach). This way refcounting is kept in the Path reconciler and we can even avoid some code duplication across reconcilers.